core/src/main/java/com/linecorp/armeria/common/stream/PathStreamMessage.java

```diff
@@ -0,0 +1,324 @@
+            this.downstream = downstream;
+            this.executor = executor;
+            this.bufferSize = bufferSize;
+            this.notifyCancellation = notifyCancellation;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                cancel();
+                downstream.onError(
+                        new IllegalArgumentException("Rule §3.9 violated: non-positive subscription requests " +
+                                                     "are forbidden."));
+            } else {
+                request0(n);
+            }
+        }
+
+        private void request0(long n) {
+            final long requested = this.requested;
+            if (requested == Long.MAX_VALUE) {
+                return;
+            }
+            if (n == Long.MAX_VALUE) {
+                this.requested = Long.MAX_VALUE;
+            } else {
+                this.requested = LongMath.saturatedAdd(requested, n);
+            }
+
+            if (requested > 0) {
+                // PathSubscription is reading a file.
+                // New requests will be handled by 'completed(Integer, ByteBuf)'.
+                return;
+            }
+
+            read();
+        }
+
+        private void read() {
+            if (!reading && !closed && requested > 0) {
+                requested--;
+                reading = true;
+                final ByteBuf buffer = alloc.buffer(bufferSize);
+                fileChannel.read(buffer.nioBuffer(0, bufferSize), position, buffer, this);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            if (!reading) {
+                maybeCloseFileChannel();
+            }
+
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (notifyCancellation) {
+                downstream.onError(cause);
+            }
+            completionFuture.completeExceptionally(cause);
+            downstream = NoopSubscriber.get();
+        }
+
+        @Override
+        public void completed(Integer result, ByteBuf byteBuf) {
+            executor.execute(() -> {
+                if (closed) {
+                    ReferenceCountUtil.release(byteBuf);
+                    maybeCloseFileChannel();
+                } else {
+                    if (result > -1) {
+                        position += result;
+                        byteBuf.writerIndex(result);
+                        downstream.onNext(HttpData.wrap(byteBuf));
+                        reading = false;
+                        read();
+                    } else {
+                        maybeCloseFileChannel();
```

Comment:
I think we need `ReferenceCountUtil.release(byteBuf);`

Reply:
Oops...